accepted/0000-dependency-selector-syntax.md

```diff
@@ -0,0 +1,422 @@
+    "description": "",
+    "homepage": "",
+    "bugs": {},
+    "author": {},
+    "license": {},
+    "funding": {},
+    "files": [],
+    "main": "",
+    "browser": "",
+    "bin": {},
+    "man": [],
+    "directories": {},
+    "repository": {},
+    "scripts": {},
+    "config": {},
+    "dependencies": {},
+    "devDependencies": {},
+    "optionalDependencies": {},
+    "bundledDependencies": {},
+    "peerDependencies": {},
+    "peerDependenciesMeta": {},
+    "engines": {},
+    "os": [],
+    "cpu": [],
+    "workspaces": {},
+    "keywords": [],
+    "ancenstry": "",
+    "path": "",
+    "realpath": "",
+    "parent": "",
+    "vulnerabilities": [],
+    "cwe": []
+  },
+  ...
+```
+
+#### Usage
+
+```bash
+npm query ":root > .workspace > *"  # get all workspace direct deps
+```
+
+### Extended Example Queries & Use Cases
+
+```stylus
+// all deps
+*
+
+// all direct deps
+:root > *
+
+// direct production deps
+:root > .prod
+
+// direct development deps
+:root > .dev 
+
+// any peer dep of a direct deps
+:root > * > .peer  
+  
+// any workspace dep
+.workspace
+
+// all workspaces that depend on another workspace
+.workspace > .workspace
+    
+// all workspaces that have peer deps
+.workspace:has(*.peer)
+
+// any dep named "lodash"
+// equivalent to [name="lodash"]
+#lodash 
+  
+// any deps named "lodash" & within semver range ^"1.2.3"
+#lodash@^1.2.3
+// equivalent to...
+[name="lodash"]:semver(^1.2.3)
+
+// get the hoisted node for a given semver range
+#lodash@^1.2.3:not(:deduped)
+
+// querying deps with a specific version
+#lodash@2.1.5
+// equivalent to...
+[name="lodash"][version="2.1.5"]
+  
+// all deps living alongside vulnerable deps
+*:vulnerable ~ *:not(:vulnerable)
+
+// has any deps
+*:has(*)
+
+// has any vulnerable deps
+*:has(*:vulnerable)
+
+// deps with no other deps (ie. "leaf" nodes)
+*:empty
+
+// all vulnerable deps that aren't dev deps & that aren't vulnerable to CWE-1333
+*:vulnerable:not(.dev:cwe(1333))
+  
+// manually querying git dependencies
+*[repository^="github:"],
+*[repository^="git:"],
+*[repository^="https://github.com"],
+*[repository^="http://github.com"],
+*[repository^="https://github.com"],
+*[repository^="+git:..."]
+  
+// querying for all git dependencies
+*:type(git)
+
+// find all references to "install" scripts
+*[scripts=install], 
+*[scripts=postinstall],
+*[scripts=preinstall]
+
+// get production dependencies that aren't also dev deps
+.prod:not(.dev)
+    
+// get dependencies with specific licenses
+*[license="MIT"], *[license="ISC"]
+
+// find all packages that have @ruyadorno as a contributor
+*:contributors([email=ruyadorno@github.com])
+```
+
+## Next Steps
+
+### Command Mapping to `query-output`
+
+Previous commands with similar behaivours will now be able to utilize `Aborist` `Node.querySelectorAll()` under-the-hood & will fast-follow the `npm query` implementation.
+
+#### `npm list`
+
+```bash
+npm list # equivalent to...
+npm query ":root > *"
```

Comment:
`npm ls` does validation of the graph, which `npm query` wouldn't do - do you mean that *some* of the output is equivalent to that?

Reply:
No. You're right, it's not 100% equivalent given that the above wouldn't throw on error.